# Analisando os algoritmos de busca aplicados ao problema da Mochila Binária

Esse notebook tem como objetivo analisar os dados obtidos ao rodarmos os algoritmos de busca em 800 casos de teste gerados de forma pseudo-aleatória.

## Importando os dados dos testes

Aqui simplesmente adicionamos as bibliotecas que serão usadas e importamos o .csv que conteḿ os resultados de cada um dos testes.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

sns.set()

import warnings
warnings.filterwarnings('ignore')

In [2]:
results = pd.read_csv("out.csv")
results.rename({" size": "Tamanho do Problema", "time":"Tempo (ms)", "count":"N° de Passos"}, inplace=True, axis=1)
results

case_test  Tamanho do Problema      search_type  N° de Passos  \
0       tests/0-test.txt                    4     Blind Search            64   
1       tests/0-test.txt                    4  Best-Fit Search             4   
2       tests/0-test.txt                    4          B&B DFS             1   
3       tests/0-test.txt                    4          B&B BFS             1   
4       tests/1-test.txt                    8     Blind Search          2048   
5       tests/1-test.txt                    8  Best-Fit Search             8   
6       tests/1-test.txt                    8          B&B DFS             1   
7       tests/1-test.txt                    8          B&B BFS             1   
8       tests/2-test.txt                   10     Blind Search         10240   
9       tests/2-test.txt                   10  Best-Fit Search            10   
10      tests/2-test.txt                   10          B&B DFS            23   
11      tests/2-test.txt                   10          B&B BFS            21   
12      tests/3-test.txt                    5     Blind Search           160   
13      tests/3-test.txt                    5  Best-Fit Search             5   
14      tests/3-test.txt                    5          B&B DFS             5   
15      tests/3-test.txt                    5          B&B BFS             3   
16      tests/4-test.txt                    7     Blind Search           896   
17      tests/4-test.txt                    7  Best-Fit Search             7   
18      tests/4-test.txt                    7          B&B DFS            21   
19      tests/4-test.txt                    7          B&B BFS            21   
20      tests/5-test.txt                    1     Blind Search             2   
21      tests/5-test.txt                    1  Best-Fit Search             1   
22      tests/5-test.txt                    1          B&B DFS             1   
23      tests/5-test.txt                    1          B&B BFS             1   
24      tests/6-test.txt                    1     Blind Search             2   
25      tests/6-test.txt                    1  Best-Fit Search             1   
26      tests/6-test.txt                    1          B&B DFS             3   
27      tests/6-test.txt                    1          B&B BFS             3   
28      tests/7-test.txt                    5     Blind Search           160   
29      tests/7-test.txt                    5  Best-Fit Search             5   
...                  ...                  ...              ...           ...   
3174  tests/793-test.txt                   13          B&B DFS            21   
3175  tests/793-test.txt                   13          B&B BFS           459   
3176  tests/794-test.txt                   24     Blind Search     402653184   
3177  tests/794-test.txt                   24  Best-Fit Search            24   
3178  tests/794-test.txt                   24          B&B DFS            55   
3179  tests/794-test.txt                   24          B&B BFS         26055   
3180  tests/795-test.txt                    9     Blind Search          4608   
3181  tests/795-test.txt                    9  Best-Fit Search             9   
3182  tests/795-test.txt                    9          B&B DFS            11   
3183  tests/795-test.txt                    9          B&B BFS            31   
3184  tests/796-test.txt                   18     Blind Search       4718592   
3185  tests/796-test.txt                   18  Best-Fit Search            18   
3186  tests/796-test.txt                   18          B&B DFS            47   
3187  tests/796-test.txt                   18          B&B BFS          2827   
3188  tests/797-test.txt                   27     Blind Search    -671088640   
3189  tests/797-test.txt                   27  Best-Fit Search            27   
3190  tests/797-test.txt                   27          B&B DFS            63   
3191  tests/797-test.txt                   27          B&B BFS          2697   
3192  tests/798

Agora, nós vamos dividir os resultados dos testes por cada um dos algoritmos.

In [3]:
blind_search = results[results[" search_type"] == "Blind Search"]
best_fit = results[results[" search_type"] == "Best-Fit Search"]
BB_DFS = results[results[" search_type"] == "B&B DFS"]
BB_BFS = results[results[" search_type"] == "B&B BFS"]

## Analisando o Blind Search

Agora, vamos analisar os resultados do Blind Search. O primeiro passo é verificar valores que possam ter gerado algum erro, pois sabemos que em alguns casos o número de passos estourou o tamanho limite da variável de contagem. Para esses casos, iremos considerar count como infinito.

In [4]:
blind_search.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 801 entries, 0 to 3200
Data columns (total 7 columns):
case_test              801 non-null object
Tamanho do Problema    801 non-null int64
 search_type           801 non-null object
N° de Passos           801 non-null int64
Tempo (ms)             801 non-null float64
ids                    732 non-null object
right_answer           732 non-null object
dtypes: float64(1), int64(2), object(4)
memory usage: 50.1+ KB


Podemos ver que em (801 - 732) casos, o caso de teste não apresentava solução. 

Vamos cuidar agora da variável count.

In [5]:
blind_search["N° de Passos"][blind_search["N° de Passos"] < 0] = np.inf

In [6]:
blind_search.describe()

Tamanho do Problema  N° de Passos    Tempo (ms)
count           801.000000         801.0    801.000000
mean             16.942572           inf   3024.689670
std               5.881683           NaN   7747.547988
min               1.000000           2.0      0.007000
25%              12.000000       49152.0      0.600000
50%              17.000000     2228224.0     22.983000
75%              22.000000    92274688.0    884.677000
max              27.000000           inf  43960.300000

### Análise do tempo de execução

Vamos verificar como o tempo varia em relação ao tamanho do problema

In [7]:
plt.figure(figsize=(15,8))
plt.title("Tempo de Execução do Blind Search", fontdict={"fontsize": 20})
sns.scatterplot(x = blind_search["Tamanho do Problema"], y = blind_search["Tempo (ms)"])
plt.savefig("graficos/blind_tempo")

ValueError: matplotlib display text must have all code points < 128 or use Unicode strings

Error in callback <function post_execute at 0x7f0aaed0c500> (for post_execute):


ValueError: matplotlib display text must have all code points < 128 or use Unicode strings

ValueError: matplotlib display text must have all code points < 128 or use Unicode strings

<Figure size 1080x576 with 1 Axes>

### Análise do número de passos

In [ ]:
plt.figure(figsize=(15,8))
plt.title("Número de Passos no Blind Search", fontdict={"fontsize": 20})
sns.scatterplot(x = blind_search["Tamanho do Problema"], y = blind_search["N° de Passos"])
plt.savefig("graficos/blind_passos")

## Analisando o Best-Fit

In [ ]:
best_fit.info()

In [ ]:
best_fit.describe()

Como podemos ver, o Best-Fit não teve problema de estourar o número de passos.

### Análise do Tempo de Execução

In [ ]:
plt.figure(figsize=(15,8))
plt.title("Tempo de Execução do Best-Fit", fontdict={"fontsize": 20})
sns.scatterplot(x = best_fit["Tamanho do Problema"], y = best_fit["Tempo (ms)"])
plt.savefig("graficos/best_tempo")

### Análise do Número de Passos

In [ ]:
plt.figure(figsize=(15,8))
plt.title("Número de Passos no Best-Fit", fontdict={"fontsize": 20})
sns.scatterplot(x = best_fit["Tamanho do Problema"], y = best_fit["N° de Passos"])
plt.savefig("graficos/best_passos")

## Analisando o Branch and Bound DFS

In [ ]:
BB_DFS.info()

In [ ]:
BB_DFS.describe()

### Análise do tempo

In [ ]:
plt.figure(figsize=(15,8))
plt.title("Tempo de Execução do BB DFS", fontdict={"fontsize": 20})
sns.scatterplot(x = BB_DFS["Tamanho do Problema"], y = BB_DFS["Tempo (ms)"])
plt.savefig("graficos/BB_DFS_tempo")

### Análise do Número de Passos

In [ ]:
plt.figure(figsize=(15,8))
plt.title("Número de Passos no BB DFS", fontdict={"fontsize": 20})
sns.scatterplot(x = BB_DFS["Tamanho do Problema"], y = BB_DFS["N° de Passos"])
plt.savefig("graficos/BB_DFS_passos")

## Análise do Branch and Bound BFS

In [ ]:
BB_BFS.info()

In [ ]:
BB_BFS.describe()

### Analisando o tempo

In [ ]:
plt.figure(figsize=(15,8))
plt.title("Tempo de Execução do BB BFS", fontdict={"fontsize": 20})
sns.scatterplot(x = BB_BFS["Tamanho do Problema"], y = BB_BFS["Tempo (ms)"])
plt.savefig("graficos/BB_BFS_tempo")

### Analisando o número de passos

In [ ]:
plt.figure(figsize=(15,8))
plt.title("Número de Passos no BB BFS", fontdict={"fontsize": 20})
sns.scatterplot(x = BB_BFS["Tamanho do Problema"], y = BB_BFS["N° de Passos"])
plt.savefig("graficos/BB_BFS_passos")

## Fazendo a análise de número de ótimos do Best_Fit

In [ ]:
best_fit["isGreat"] = False

In [ ]:
def verify_great(df, df_name, img_name):

    cont = 0
    for index, row in df.iterrows():
        resultado = row["ids"]
        verdadeiro = row["right_answer"]

        if(resultado is np.nan or verdadeiro is np.nan):
            df.at[index,"isGreat"] = True
            cont += 1
            continue

        resultado = resultado.split(" ")
        verdadeiro = verdadeiro.split(" ")


        resultado.remove("")
        verdadeiro.remove("")

        resultado = list(map(int,resultado))
        verdadeiro = list(map(int,verdadeiro))

        resultado.sort()
        verdadeiro.sort()

        if(resultado == verdadeiro):
            df.at[index,"isGreat"] = True
            cont += 1
    
    plt.figure(figsize=(8,8))
    plt.title(str(df_name), fontdict={"fontsize": 20})

    plt.pie(df["isGreat"].value_counts(), autopct="%1.1f%%", textprops={'color': "w", "fontsize":20})
    plt.savefig("graficos/{}".format(img_name))

In [ ]:
verify_great(best_fit, "Porcentagem de Ótimos do Best Fit", "otimo_best")

In [ ]:
blind_search["isGreat"] = False
verify_great(blind_search, "Porcentagem de Ótimos do Blind Search", "otimo_blind")

In [ ]:
BB_BFS["isGreat"] = False
verify_great(BB_BFS, "Porcentagem de Ótimos do B&B BFS", "otimo_BB_BFS")

In [ ]:
BB_DFS["isGreat"] = False
verify_great(BB_DFS, "Porcentagem de Ótimos do B&B DFS", "otimo_BB_DFS")